### PR TITLE
PowerSubsystem: Add PowerSupplyMetrics

### PR DESCRIPTION
**Pull request by smccarney**
Add support for the PowerSupplyMetrics Redfish schema.

Add the Metrics property to the Redfish PowerSupply output.  This property contains the URI for the power supply metrics.

Implement basic support for the PowerSupplyMetrics schema.  Initially support only power supply input power history.  Input history is not currently defined in the Redfish specification, and it is defined in OpenPower D-Bus interfaces, so it is stored in an IBM OEM property.

Modify the XML and JSON PowerSupplyMetrics schemas to define the IBM OEM property.

Tested:
* Passed the Redfish Validator
* Verified power supply input history displayed for chassis PSUs
* Verified power supply input history not displayed for I/O drawer PSUs
* For detailed test plan, see https://gist.github.com/smccarney/b3df3fc5142dbff0b0c502654ec1a5ac

Example Output:
```
$ curl -k https://${bmc}/redfish/v1/Chassis/chassis/PowerSubsystem/PowerSupplies/powersupply0 {
  "@odata.id": "/redfish/v1/Chassis/chassis/PowerSubsystem/PowerSupplies/powersupply0",
  "@odata.type": "#PowerSupply.v1_5_0.PowerSupply",
  "EfficiencyRatings": [
    {
      "EfficiencyPercent": 90
    }
  ],
  "FirmwareVersion": "423243324132",
  "Id": "powersupply0",
  "Location": {
    "PartLocation": {
      "ServiceLabel": "U78DA.ND0.WZS004K-E0"
    }
  },
  "LocationIndicatorActive": false,
  "Manufacturer": "",
  "Metrics": {
    "@odata.id": "/redfish/v1/Chassis/chassis/PowerSubsystem/PowerSupplies/powersupply0/Metrics"
  },
  "Model": "51E9",
  "Name": "powersupply0",
  "PartNumber": "03FP379",
  "SerialNumber": "YL30KY1450M3",
  "SparePartNumber": "03FP378",
  "Status": {
    "Health": "OK",
    "State": "Enabled"
  }
}

$ curl -k https://${bmc}/redfish/v1/Chassis/chassis/PowerSubsystem/PowerSupplies/powersupply0/Metrics {
  "@odata.id": "/redfish/v1/Chassis/chassis/PowerSubsystem/PowerSupplies/powersupply0/Metrics",
  "@odata.type": "#PowerSupplyMetrics.v1_0_1.PowerSupplyMetrics",
  "Id": "Metrics",
  "Name": "Metrics for powersupply0",
  "Oem": {
    "@odata.type": "#OemPowerSupplyMetrics.v1_0_0.Oem",
    "IBM": {
      "@odata.type": "#OemPowerSupplyMetrics.v1_0_0.IBM",
      "InputPowerHistoryItems": [
        {
          "Average": 265.0,
          "Date": "2023-04-25T04:40:56.530+00:00",
          "Maximum": 266.0
        },
        {
          "Average": 265.0,
          "Date": "2023-04-25T04:40:26.499+00:00",
          "Maximum": 266.0
        },
        {
          "Average": 266.0,
          "Date": "2023-04-25T04:39:56.468+00:00",
          "Maximum": 276.0
        }
      ]
    }
  }
}
```

Change-Id: I1e9523a87d016dde499a98c1262cea2b3a4d6557

### CODE REVIEW

File: redfish-core/include/utils/power_supply_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    std::string associationPath = validPowerSupplyPath + "/input_history";
+    dbus::utility::getAssociationEndPoints(
+        associationPath, [asyncResp, callback{std::move(callback)}](
+                             const boost::system::error_code& ec,
+                             const dbus::utility::MapperEndPoints& endpoints) {
+            if (ec)
+            {
+                if (ec.value() != EBADR)
+                {
+                    messages::internalError(asyncResp->res);
```
Comment (gtmills) [inline redfish-core/include/utils/power_supply_utils.hpp:37]: A lot times we do  LOG_ERROR with the ec value 
``` BMCWEB_LOG_ERROR << ``` when we have an internalError, it helps with debug. Oh well, I feel these aren't that important
 E.g. 
 https://github.com/ibm-openbmc/bmcweb/blob/1050/redfish-core/lib/account_service.hpp#L409

File: redfish-core/include/utils/power_supply_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    std::string powerPath = validChassisPath + "/powered_by";
+    dbus::utility::getAssociationEndPoints(
+        powerPath, [asyncResp, powerSupplyId, callback{std::move(callback)}](
+                       const boost::system::error_code& ec,
+                       const dbus::utility::MapperEndPoints& endpoints) {
+            if (ec)
+            {
+                if (ec.value() != EBADR)
+                {
+                    messages::internalError(asyncResp->res);
```
Comment (gtmills) [inline redfish-core/include/utils/power_supply_utils.hpp:75]: A lot times we do LOG_ERROR with the ec value

File: redfish-core/lib/power_supply_metrics.hpp [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    getInputHistoryServiceAndInterface(
+        asyncResp, historyPaths.front(),
+        [asyncResp, historyPaths](const std::string& service,
+                                  const std::string& interface) mutable {
+        // Get all properties from the first history path
+        sdbusplus::asio::getAllProperties(
+            *crow::connections::systemBus, service, historyPaths.front(),
+            interface,
+            [asyncResp, historyPaths,
+             interface](const boost::system::error_code ec,
```
Comment (gtmills) [inline redfish-core/lib/power_supply_metrics.hpp:153]: Pass error_code by reference please 
https://gerrit.openbmc.org/c/openbmc/bmcweb/+/61013/6/http/http_server.hpp

File: redfish-core/lib/power_supply_metrics.hpp [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+                        const dbus::utility::DBusPropertiesMap&
+                            propertiesList) mutable {
+            if (ec)
+            {
+                messages::internalError(asyncResp->res);
```
Comment (gtmills) [inline redfish-core/lib/power_supply_metrics.hpp:158]: LOG_ERROR if you think it would help

File: redfish-core/lib/power_supply_metrics.hpp [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+}
+
+} // namespace power_supply_metrics
+
+inline void requestRoutesPowerSupplyMetrics(App& app)
+{
+    BMCWEB_ROUTE(
+        app,
+        "/redfish/v1/Chassis/<str>/PowerSubsystem/PowerSupplies/<str>/Metrics")
+        .privileges(redfish::privileges::headPowerSupplyMetrics)
```
Comment (gtmills) [inline redfish-core/lib/power_supply_metrics.hpp:276]: Thanks for updating to the new requirements/new style from upstream!

File: redfish-core/lib/power_supply_metrics.hpp [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    // Get Scale and Values properties
+    int64_t scale{0};
+    std::vector<std::tuple<uint64_t, int64_t>> values{};
+    const bool success = sdbusplus::unpackPropertiesNoThrow(
+        dbus_utils::UnpackErrorPrinter(), propertiesList, "Scale", scale,
+        "Values", values);
+
+    if (!success)
+    {
+        messages::internalError(asyncResp->res);
```
Comment (gtmills) [inline redfish-core/lib/power_supply_metrics.hpp:62]: LOG_ERROR if you think it would help debug

**Review by gtmills — APPROVED** [review]
Really to move away from this OEM for a future release. :) 
Two places to get telemetry data :( 
Code looks good; few minor suggestions.

**Comment by smccarney** [comment]
I pushed up a second commit to implement Gunnar's suggestions and to resolve a CI formatting error.

**Comment by gtmills** [comment]
2nd commit looks good. 
https://github.com/ibm-openbmc/bmcweb/pull/677/commits/d94cc6e9edc1a227d979dbf9e6b6fbdc0f2fdf66

**Pull request merged**